gcc/rust/ast/rust-ast-dump.h

```diff
@@ -80,6 +80,18 @@ class Dump : public ASTVisitor
    */
   template <typename T> void visit (std::unique_ptr<T> &node);
 
+  /**
+   * @see visit<std::unique_ptr<T>>
+   *
+   * This ugly template trickery is necessary to make the visit overload work
+   * for both references to AST nodes/properties and unique-pointer ones.
+   * Otherwise this overload would be always preferred.
+   */
+  template <typename T1, typename T2,
+	    typename = std::enable_if<
+	      !std::is_convertible<T1, std::unique_ptr<T2>>::value>>
+  void visit (T1 &node);
```

Comment:
I'm not sure I understand why this is a problem or if we need to do that (but that might just be my C++ skills lacking).
In another part of the dump, we have the following code:
```c++
  /**
   * Use the AST Dump as a debugging tool
   */
  template <typename T> static void debug (T &instance)
  {
    auto dump = Dump (std::cerr);

    std::cerr << '\n';
    instance.accept_vis (dump);
    std::cerr << '\n';
  }
  template <typename T> static void debug (std::unique_ptr<T> &instance)
  {
    debug (*instance);
  }
```

Doesn't this effectively achieve the same behavior without the template condition magic?

Reply:
I will try that. It would be great to get rid of that magic. I wonder if calling the other overload does the trick, because this is pretty much what I started with except that one was not calling the other, but both called accept_vis.